Add getAdvertistingPostByIdOrImportEntity support to mindful package

B77Mills opened the pull request:
- Add support for getAdvertisingPostByIdOrImportEntity to the api client.  
 - Centralize fragments, utilzs and route logic
 - Update marko-web mindful service with updated middleware calls for getAdvertisingPostByIdOrImportEntity
 
 http://www-wattagnet.dev.parameter1.com:9926/brand-insights/zoetis/vaccination-success-depends-on-crew-training/67094ddcc2bbee3e1b1af57d?pubid=644d5a33b225a200013c2dd2&utm_source=NativeX&utm_medium=banner&utm_campaign=67129a46b42e363e8d7c9ae3&utm_term=646792140ec6330001a96883&utm_content=
 
<img width="45%" alt="Screenshot 2024-11-04 at 10 53 32 AM" src="https://github.com/user-attachments/assets/8630a6fd-38b9-44fd-a7f4-995288219255">

 
 
**start-server.js cal:**
 ```js
 const MindfulMarkoWebService = require('@parameter1/base-cms-mindful/marko-web/middleware/service');

const { namespace } = getAsObject(options, 'siteConfig.mindful');
app.use(MindfulMarkoWebService({ namespace }));

```
 
 
 **The following routes can be added withing the site repos with the following route calls:** 
 ```js
 // include at the top of routes:
 const advertisingPostByIdOrImportEntity = require('@parameter1/base-cms-mindful/marko-web/middleware/advertising-post-by-id-or-import-entity');
const advertisingPostTemplate = require('../templates/content/advertising-post');

// withing the routes list 
  getAdvertisingPostAsNativeStory(app, {
    route: '/brand-insights/:section/:slug/:id',
    tenant: 'watt',
    provider: 'native-x',
    template: advertisingPostTemplate,
  });
 ```
 
 **Optionally you can also set a fragment to override the following:**
 ```js
 
 const advertisingPostById = require('../../graphql/fragments/advertising-post-by-id');

 // withing the routes list 
  advertisingPostByIdOrImportEntity(app, {
    route: '/brand-insights/:section/:slug/:id',
    tenant: 'watt',
    provider: 'native-x',
    template: advertisingPostTemplate,
    fragment: advertisingPostById
  });
  ```
**Example Fragment**
 ```gql
 
 fragment AdvertisingPostById on AdvertisingPost {
  _id
  title {
    default
  }
  teaser
  body
  url
  publishedDay
  statusEdge {
    node {
      _id
      label
    }
  }
  featuredImageEdge {
    _id
    node {
      _id
      src {
        url
        settings {
          fpX
          fpY
        }
      }
    }
  }
  companyEdge {
    _id
    node {
      _id
      name {
        default
      }
      logoEdge {
        _id
        node {
          _id
          src {
            url
            settings {
              fpX
              fpY
            }
          }
        }
      }
    }
  }
}
```

B77Mills committed:
Initial restructure of geAdvertiserPostById

 - Add support for getAdvertiserPostByIdOrImportEntity

B77Mills committed:
Fix formatting

B77Mills committed:
cleanup advertisingPostById route params

B77Mills committed:
more route cleanup

B77Mills committed:
more formatting cleanup

B77Mills committed:
Fix formatting

B77Mills committed:
update fragment name

B77Mills committed:
Kidding make the name match the fragment name

B77Mills committed:
call and spread shared post by id fragment

B77Mills committed:
Cleanup and remove un-used gql exports.

zarathustra323 commented on packages/mindful/api-client.js:
```diff
@@ -51,6 +51,41 @@ class MindfulApiClient {
+      _id,
+      provider,
+      tenant,
+      type = 'post',
```
`type` shouldn't have a default value set

zarathustra323 commented on packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js:
```diff
@@ -0,0 +1,27 @@
+const { asyncRoute } = require('@parameter1/base-cms-utils');
```
There only needs to be one marko web middleware... and that's this one. Rename this to `advertising-post-as-native-story.js` and then delete the `advertising-post-by-id.js`

zarathustra323 commented on packages/mindful/marko-web/middleware/advertising-post-by-id.js:
```diff
@@ -0,0 +1,20 @@
+const { asyncRoute } = require('@parameter1/base-cms-utils');
+const { get } = require('@parameter1/base-cms-object-path');
```
This file can be removed

zarathustra323 commented on packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js:
```diff
@@ -0,0 +1,27 @@
+    const { service } = req.mindful;
+    const { id: _id } = req.params;
+    const { namespace } = tenant;
+    const result = await service.getAdvertisingPostByIdOrImportEntity({
```
Rename the service method to `getAdvertisingPostAsNativeStory` and move the conversion to that method... see below.

In other words this should become
```js
const story = service.getAdvertisingPostAsNativeStory({ _id, preview: Boolean(req.query.preview), tenant });
res.marko(template, { story });
```

zarathustra323 commented on packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js:
```diff
@@ -0,0 +1,27 @@
+    const result = await service.getAdvertisingPostByIdOrImportEntity({
+      _id,
+      provider,
+      tenant: namespace,
```
The `namespace` is not the tenant in this case... the tenant is the `native-x` tenant key, e.g. `watt`, not `watt/default`

zarathustra323 commented on packages/mindful/marko-web/service.js:
```diff
@@ -20,5 +20,26 @@ class MindfulMarkoWebService {
+   * @param {string} externalId
+   * @returns {Promise}
+   */
+  async getAdvertisingPostByIdOrImportEntity({
```
Change this to `getAdvertisingPostAsNativeStory` with `_id`, `tenant` and `preview` as args. The provider can be hardset to `native-x` and the `type` hard set to `story`.

Use the `convertAdvertisingPostToNativeStory` directly in this method.

zarathustra323 commented on packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js:
```diff
@@ -0,0 +1,27 @@
+      provider,
+      tenant: namespace,
+    }, fragment);
+    const story = service.convertAdvertisingPostToNativeStory({
```
this middleware should also return a 404 page ithe story cannot be found.

zarathustra323 commented on packages/mindful/marko-web/middleware/advertising-post-by-id-or-import-entity.js:
```diff
@@ -0,0 +1,27 @@
+      advertisingPost: getAsObject(result, 'data.advertisingPostByIdOrImportEntity'),
+      preview: Boolean(req.query.preview),
+    });
+    res.marko(template, { story });
```
There used to be handling for redirecting the page is the slug is incorrectly entered or if it changes... but I don't see that happening here?

B77Mills committed:
Consolidate postById & orImportEntity into PostAsNativeStory instead

convert  getAdvertisingPostById & getAdvertisingPostByIdOrImportEntity into getAdvertsitingPostAsNativeStory and assume to use the later api.

B77Mills committed:
clean utilz

B77Mills committed:
return undefinded for invalid post and throw 404

B77Mills committed:
Add redirect of trailing /

B77Mills committed:
cleanup

B77Mills committed:
return null for consistant return

B77Mills committed:
redirect random miss match patterns to storyURL

B77Mills committed:
Move convert to middleware service not the api

B77Mills committed:
Allow for canonical-url to be passed threw to component

This will come from the default layout for advertising post

Will need org repos to also have their advertisting post template add the following call
`<theme-default-page title=story.title description=story.teaser canonical-url=story.url>`

B77Mills committed:
cleanup and remove convert ad in api

B77Mills committed:
return null  :)

B77Mills committed:
add missing ;

B77Mills committed:
move getAdPostAsNative to middlewareservice & leave api generic

zarathustra323 commented on packages/mindful/api-client.js:
```diff
@@ -51,6 +51,43 @@ class MindfulApiClient {
+    {
+      _id,
+      provider,
+      tenant = 'native-x',
```
Remove default values from `tenant` and `type` params

zarathustra323 commented on packages/mindful/marko-web/middleware/get-advertising-post-as-native-story.js:
```diff
@@ -0,0 +1,42 @@
+  route,
+  template,
+  tenant,
+  provider,
```
Do not pass the `provider` param

zarathustra323 commented on packages/mindful/marko-web/middleware/get-advertising-post-as-native-story.js:
```diff
@@ -0,0 +1,42 @@
+      _id,
+      preview: Boolean(req.query.preview),
+      tenant,
+      provider,
```
Do not pass the `provider` param

zarathustra323 commented on packages/mindful/marko-web/service.js:
```diff
@@ -16,9 +17,22 @@ class MindfulMarkoWebService {
-    return post;
+  async getAdvertisingPostAsNativeStory({
+    _id,
+    preview,
```
Remove the `provider` and `type` params.. only allow `tenant`

zarathustra323 commented on packages/mindful/marko-web/service.js:
```diff
@@ -16,9 +17,22 @@ class MindfulMarkoWebService {
+    const response = await this.client.getAdvertisingPostByIdOrImportEntity({
+      _id,
+      preview,
+      provider,
```
Hard set provider and type to:
```js
provider: 'native-x',
type: 'story',
```

B77Mills committed:
Fix prop setting

B77Mills committed:
call right service function

B77Mills committed:
just use queryName

B77Mills committed:
move prop up a level

B77Mills committed:
remove @todo

B77Mills commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add monorail/mindful package to replace with native story middleware with Mindful equivalent](https://trello.com/c/zU4rTgw1/547-add-monorail-mindful-package-to-replace-with-native-story-middleware-with-mindful-equivalent)
